arm-service-map/2015-11-01-preview/swagger/arm-service-map.json

```diff
@@ -0,0 +1,2508 @@
+          },
+          {
+            "name": "endTime",
+            "in": "query",
+            "required": false,
+            "type": "string",
+            "format": "date-time",
+            "description": "UTC date and time specifying the end time of an interval. When not specified the service uses DateTime.UtcNow"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "The operation completed successfully.",
+            "schema": {
+              "$ref": "#/definitions/MachinesSummary"
+            }
+          },
+          "default": {
+            "description": "An error occurred while processing the request. See the error.code parameter to identify the specific error.",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponse"
+            }
+          }
+        }
+      }
+    }
+  },
+  "definitions": {
+    "Resource": {
+      "description": "Base for all ARM-adressable resources",
+      "properties": {
+        "id": {
+          "type": "string",
+          "description": "Resource identifier."
+        },
+        "type": {
+          "type": "string",
+          "description": "Resource type."
+        },
+        "name": {
+          "type": "string",
+          "description": "Resource name."
+        },
+        "etag": {
+          "type": "string",
+          "description": "Resource ETAG."
+        },
+        "kind": {
+          "type": "string",
+          "description": "Additional resource type qualifier."
+        }
+      },
+      "x-ms-external": true,
+      "x-ms-azure-resource": true,
+      "required": [
+        "id",
+        "type",
+        "name"
+      ]
+    },
+    "ResourceReference": {
+      "description": "Represents a reference to another resource.",
+      "discriminator": "kind",
+      "properties": {
+        "id": {
+          "type": "string",
+          "description": "Resource URI."
+        },
+        "type": {
+          "type": "string",
+          "description": "Additional resource type qualifier.  Always present when retrieving data from the service.  Optional when sending data to the service."
+        },
+        "name": {
+          "type": "string",
+          "description": "Resource name. Always present when retrieving data from the service.  Optional when sending data to the service."
+        },
+        "kind": {
+          "type": "string",
+          "description": "Specifies the sub-class of the reference."
+        }
+      },
+      "required": [
+        "id"
+      ],
+      "x-ms-azure-resource": true
+    },
+    "ResourceOnMachineReference": {
+      "description": "Reference to a resource on a machine.",
+      "allOf": [
+        {
+          "$ref": "#/definitions/ResourceReference"
+        }
+      ],
+      "x-ms-discriminator-value": "ref:onmachine",
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
```

Comment:
I don't think I follow: the service returns a ResourceReference. The client side library code defines all operations in terms of ResourceReference. When deserializing at runtime, the "kind" field is used to determine what is being returned and an instance of the right class is instantiated, in this case "ResourceOnMachineReference". Client code operates on it by default as a ResourceReference and only when desired can typecast it to "ResourceOnMachineReference": this will be an explicit typecast by the user of the library to obtain arguments needed to retrieve the referenced resources using the library. Similarly if the api always returns an instance of ResourceOnMachineReference, then the contract would express that and the client code will treat it as ResourceOnMachineReference.

Note that the base class ResourceReferences does not have a nested "properties" element, so any flattening applies only on instances of ResourceOnMachineReference.